src/db.c

```diff
@@ -152,13 +143,13 @@ robj *lookupKeyRead(redisDb *db, robj *key) {
 }
 
 /* Lookup a key for write operations, and as a side effect, if needed, expires
- * the key if its TTL is reached.
+ * the key if its TTL is reached. It's equivalent to lookupKey() with the
+ * LOOKUP_WRITE flag added.
  *
  * Returns the linked value object if the key exists or NULL if the key
  * does not exist in the specified DB. */
 robj *lookupKeyWriteWithFlags(redisDb *db, robj *key, int flags) {
-    expireIfNeeded(db,key);
-    return lookupKey(db,key,flags);
+    return lookupKey(db, key, flags | LOOKUP_WRITE);
```

Comment:
here the flag with `LOOKUP_WRITE` would cause delete the expired key on replica, but that may lead data inconsistency if the replication delayed, for example:

we have a key (e.g. `hash`) with ttl 10, and apply a `hset` command on master, and then before ttl 10 we apply `persist` command on it. but replica receives the `hset` and `persist` after ttl 10, the `lookupKeyWrite` would delete the key.

Reply:
it was not supposed to do that. either you found an unintended bug, or we're missing something.

LOOKUP_WRITE was only supposed to prevent expiration on a replica when the command did not originate from the master client. i.e. since it must be a write command, it's only about writable-replicas.

but indeed when looking at the current code i fail to find the place where we make that decision (i.e. look at `server.current_client == server.master` or check the `CLIENT_MASTER` flag.

@zuiderkwast are we missing something? if we don't, then i wonder how come the tests didn't find that bug (i'm quite sure we have a test for it)

Reply:
Do you mean when lookupKeyWrite is called in during replication? I didn't think about this scenario. I don't know if we have tests for it. I guess it's a race condition which is hard to catch(?).

I think what has changed is this line in expireIfNeeded:

```diff
-    if (server.masterhost != NULL) return 1;
+    if (server.masterhost != NULL && !expire_on_replica) return 1;
```

Before, replicas never deleted the expired key here and now they do. Indeed I think we need to add a check for `server.current_client == server.master` or check the `CLIENT_MASTER` flag.

Reply:
> Do you mean when lookupKeyWrite is called in during replication?

It's a trivial case of INCR being executed on a volatile key on the master, and how the replica will handle that command when is propagated to it.

the old code had this check `server.current_client != server.master` which prevented the key from being evicted and let the lookup succeed despite being already expired. i didn't notice we lost it, but we must re-add it.

The test for it is not too complicated:
1. disable active expire on the master
2. create a volatile key with the value of 1 and expiration of 1 second
3. suspend the replica using SIGSTOP
4. run INCR on the master.
5. wait 1 second.
6. unpause the replica
7. wait for replication offset to sync
8. use DEBUG OBJECT to check that the key exists, and maybe extend it so we can also make sure it's still volatile.
9. or, instead, promote the replica and use TTL and GET to see that it was incremented (value is 2) and volatile.

Reply:
@zuiderkwast i see you fixed it, but please add that test too (and make sure it fails on the buggy version).
thanks.

Reply:
Test added as suggested. I have confirmed that it fails when I comment out the line `if (server.current_client == server.master) return 0;` in expireIfNeeded.